File: algorithm study/BOJ/2. Silver/1914.js
// boj 1914 하노이 탑
// recursion

const input = require("fs").readFileSync("./example.txt").toString().split("\n");
// const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");

const n = input[0].split(" ").map(Number); //1~100
let result = [];
function hanoi(n, start, via, end) {
  if (n == 1) {
    result.push([start, end]);
    return;
  }
  hanoi(n - 1, start, end, via); //n-1개를 장대1 -> 장대2
  result.push([start, end]);
  hanoi(n - 1, via, start, end); //n-1개를 장대2 -> 장대3
}

console.log((2n ** BigInt(input) - 1n).toString());
if (n > 20) return;
hanoi(n, 1, 2, 3);
console.log(result.join("\n"));

File: algorithm study/BOJ/2. Silver/1914.cpp
// boj 1914 하노이 탑
// recursion
#include <cmath>
#include <iostream>
using namespace std;

void hanoi(int n, int start, int via, int end) {
    if (n == 1) {
        cout << start << " " << end << "\n";
        return;
    }
    hanoi(n - 1, start, end, via);
    cout << start << " " << end << "\n";
    hanoi(n - 1, via, start, end);
}
int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int n;  // 1 ≤ N ≤ 100
    cin >> n;
    string total = to_string(pow(2, n));
    int idx = total.find('.');
    total = total.substr(0, idx);
    total[total.length() - 1] -= 1;
    cout << total << "\n";
    if (n <= 20) hanoi(n, 1, 2, 3);

    return 0;
}